### ISSUE TITLE
[@typescript-eslint/typedef] variableDeclaration broken with for..of loops

### ISSUE THREAD

**Issue by octogonz** [typescript-eslint/typescript-eslint#783]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/typedef": [
      "error",
      {
        "variableDeclaration": true
      },
    ],
  }
}
```

```TS
// error @typescript-eslint/typedef : expected item to have a type annotation
for (const item of ['a', 'b']) {
}
```

**Expected Result**

The error should not be reported for a for..of loop variable.

**Actual Result**

The error is reported.  The error is impossible to fix, because TypeScript does not allow type annotations here.

**Additional Info**

TSLint encountered this same problem as https://github.com/palantir/tslint/issues/743.

Their fix is here and might be relevant:  https://github.com/palantir/tslint/pull/745

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0-alpha.4` |
| `@typescript-eslint/parser`        | `1.13.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |

**Comment by octogonz** [comment]
@JoshuaKGoldberg TSLint does it like this:

```ts
    public visitVariableDeclaration(node: ts.VariableDeclaration) {
        // AFAIK, variable declarations will always have a grandparent, 
        // but check that to be on the safe side.
        // catch statements will be the parent of the variable declaration
        // for-in/for-of loops will be the gradparent of the variable declaration
        if (node.parent != null && node.parent.parent != null
                && node.parent.kind !== ts.SyntaxKind.CatchClause
                && node.parent.parent.kind !== ts.SyntaxKind.ForInStatement
                && node.parent.parent.kind !== ts.SyntaxKind.ForOfStatement) {
            this.checkTypeAnnotation("variable-declaration", node.name.getEnd(), node.type, node.name);
        }
        super.visitVariableDeclaration(node);
    }
```

**Comment by octogonz** [comment]
Here's a PR: https://github.com/typescript-eslint/typescript-eslint/pull/787
